src/core/config.c

```diff
@@ -0,0 +1,109 @@
+#include "config.h"
+#include "xdpw.h"
+#include "logger.h"
+
+#include <dictionary.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <iniparser.h>
+
+void print_config(enum LOGLEVEL loglevel, struct xdpw_config *config) {
+	logprint(loglevel, "config: Outputname  %s",config->screencast_conf.output_name);
+}
+
+// NOTE: calling finish_config won't prepare the config to be read again from config file
+// with init_config since to pointers and other values won't be reset to NULL, or 0
+void finish_config(struct xdpw_config *config) {
+	logprint(DEBUG, "config: destroying config");
+
+	// screencast
+	free(&config->screencast_conf.output_name);
+}
+
+static void getstring_from_conffile(dictionary *d, char *key, char **dest, char *fallback) {
+	if (*dest != NULL) {
+		return;
+	}
+	const char *c = iniparser_getstring(d, key, fallback);
+	if (c == NULL) {
+		return;
+	}
+	// Allow keys without value as default
+	if (strcmp(c, "") != 0) {
+		*dest = strdup(c);
+	} else {
+		*dest = fallback ? strdup(fallback) : NULL;
+	}
+}
+
+static bool file_exists(const char *path) {
+	return path && access(path, R_OK) != -1;
+}
+
+static char *config_path(char *prefix, char *filename) {
+	if (!prefix || !prefix[0] || !filename || !filename[0]) {
+		return NULL;
+	}
+
+	char *config_folder = "xdg-desktop-portal-wlr";
+
+	size_t size = 3 + strlen(prefix) + strlen(config_folder) + strlen(filename);
+	char *path = calloc(size, sizeof(char));
+	snprintf(path, size, "%s/%s/%s", prefix, config_folder, filename);
+	return path;
+}
+
+static void config_parse_file(char *configfile, struct xdpw_config *config) {
+	dictionary *d = iniparser_load(configfile);
+
+	// screencast
+	getstring_from_conffile(d, "screencast:output_name", &config->screencast_conf.output_name, NULL);
+
+	iniparser_freedict(d);
+	logprint(DEBUG, "config: config file parsed");
+	print_config(DEBUG, config);
+}
+
+static char *get_config_path(void) {
+	const char *home = getenv("HOME");
+	size_t size_fallback = 1 + strlen(home) + strlen("/.config");
+	char *config_home_fallback = calloc(size_fallback, sizeof(char));
+	snprintf(config_home_fallback, size_fallback, "%s/.config", home);
+
+	char *prefix[4];
+	prefix[0] = getenv("XDG_CONFIG_HOME");
+	prefix[1] = config_home_fallback;
+	prefix[2] = SYSCONFDIR "/xdg";
+	prefix[3] = SYSCONFDIR;
+
+	char *config[2];
+	config[0] = getenv("XDG_CURRENT_DESKTOP");
+	config[1] = "config";
+
+	for (size_t i = 0; i < 4; i++) {
+		for (size_t j = 0; j < 2; j++) {
+			char *path = config_path(prefix[i], config[j]);
+			if (!path) {
+				continue;
+			}
+			logprint(TRACE, "config: trying config file %s", path);
+			if (file_exists(path)) {
+				return path;
+			}
+			free(path);
+		}
+	}
+
+	return NULL;
+}
+
+void init_config(char *configfile, struct xdpw_config *config) {
+	if (configfile == NULL) {
+		configfile = get_config_path();
```

Comment:
Maybe we should check `configfile != NULL` at this point. `get_config_path` can return NULL.

Reply:
Yeah, but afaics libiniparser could handle NULL as a path to the config file and will insert the defaults as if this file is empty. If you want to avoid this, I could write another function setting the default values manually.

Reply:
Hm, looking at the code, it doesn't seem like it?

https://github.com/ndevilla/iniparser/blob/f858275f7f307eecba84c2f5429483f9f28007f8/src/iniparser.c#L732

Reply:
(IMHO just set the defaults in `main`)

Reply:
It happens here https://github.com/ndevilla/iniparser/blob/f858275f7f307eecba84c2f5429483f9f28007f8/src/iniparser.c#L419

About defaulting in main is, i can't detect if something is a default or was set via commandline parameter.